### DISCUSSION TITLE
Great Job! What about MIDI port?

### DISCUSSION THREAD

**Discussion by ojungleboy** [corrados/edrumulus#52] [Q&A] [answered]
First of all, great job! I am looking forward to upload it to my esp32 once it arrives.

I am planning to use a Smartwav2, its a sound module triggered by MIDI.
So I ask you, do you have plans to add a MIDI out to the schematics?

Once again, thank you!

**Comment by corrados** [comment, marked as answer]
I started designing some circuits for a MIDI in/out port in https://github.com/corrados/edrumulus/blob/main/hardware/prototype2/edrumulus.pdf some time ago. But I never bought any components for it since I do not need it. MIDI over USB works fine for me.

> So I ask you, do you have plans to add a MIDI out to the schematics?

As a result, right now I do not plan to support a MIDI out port.

**Reply by corrados** [reply]
Update: The newest [Prototype 6](https://github.com/corrados/edrumulus/tree/main/hardware/prototype6) will feature MIDI IN/OUT connectors.

**Reply by jstma** [reply]
So that Smartwav2 looks like an interesting device.  I'm running into limitations with my dtx12 and a midi sample player sounds like a great idea.  I did some research into that Smartwav2 and it has some limitations:

1. only 14 sound polyphony
2. latency of 2-6ms
3. only 4 velocity layers
4. no aux input
5. only open/close hihat

Doesn't the rpi + soundcard + drumgizmo do better?  I take it that's what @corrados is using?  He's using a usb sound card.  It seems that edrumlus sends midi to the rpi, so it should be possible to replace the edrumulus board with a board that has a midi connector.

Reading from the sd card real time is probably what limits the smartwav2 to 14 channels of polyphony.  The rpi has lots of RAM to store samples for immediate use.  The smartwav2 doesn't appear to have a codec on board, so I'm guessing the audio output is generated using the internal DAC.  Not exactly the best choice.

I like the idea of kits stored as plain wav files on an sd card.  With the right software on a computer, this could be less painful for the user.  Once my kit is set up, I don't really need to change it.  Having the ability to change the amount of reverb on the fly would be useful though.

If rpi + soundcard + drumgizmo isn't the answer, is there an alternative that is not a laptop?

**Reply by corrados** [reply]
> Doesn't the rpi + soundcard + drumgizmo do better? I take it that's what @corrados is using?

I had to tweak Drumgizmo a bit to play only a minimum number of layers to be able to run it on the Pi 3. So, the smartwav2 might even be more powerful. I have no idea what the Pi 4 can do. I expect this to perform much better.

**Comment by jstma** [comment]
I'm surprised the combination of rpi+soundcard+drumgizmo performs poorly.

The smartwav2 isn't necessarily powerful.  It's just not running an OS.  So it can respond to interrupts.  It's reading from the sd card realtime because it has no real memory to buffer it in.

The issue with Linux is that it's a multitasking OS.  It's difficult to keep it from switching away from the work you need to do.

Having said that, it's certainly possible with a real time kernel (maybe even just realtime scheduling) to prioritise the work that needs to be done.  I haven't looked into Drumgizmo's source code, but are they changing to realtime priority?  Still an rt kernel might be required.

Since the pi isn't doing anything else (ie. no network) it would be a good idea to uninstall or disable unused services.

In order to keep latency down, you'd need to do all the processing within the audio interrupt time.  Some optimization here goes a long way.  Maybe I'll have a peek.  I have an rpi3 doing nothing.  How does Drumgizmo receive midi?

**Reply by corrados** [reply]
> I'm surprised the combination of rpi+soundcard+drumgizmo performs poorly.

I didn't say that it performs poorly. You can see how it performs in [this video where I have run Drumgizmo on a Raspberry Pi 3](https://youtu.be/tBlKvivwAE4).

> Having said that, it's certainly possible with a real time kernel (maybe even just realtime scheduling) to prioritise the work that needs to be done. I haven't looked into Drumgizmo's source code, but are they changing to realtime priority? Still an rt kernel might be required.

The problem seems to be the RAM. A lot of wave forms have to be loaded and sometimes the kernel uses the SD card to load/re-load wave forms. This should be solved with the Raspberry Pi 4 which has up to 8 GB RAM.

> How does Drumgizmo receive midi?

[Here ](https://github.com/corrados/edrumulus/blob/f29c9f82ab891044d3721ca7b946d3af252f6418/tools/run_edrumulus.sh#L144)you can see how Drumgizmo is configured. The MIDI/audio is transferred with Jack Audio.
